apps/emqx/src/emqx_router_helper.erl

```diff
@@ -192,5 +122,61 @@ code_change(_OldVsn, State, _Extra) ->
 stats_fun() ->
     emqx_stats:setstat('topics.count', 'topics.max', emqx_router:stats(n_routes)).
 
+%% This is just to avoid cleaning up routes twice in case the node is down,
+%% as we can expect {mria, down}, {node, down} under normal circumstances
+%%  when a replicant node is being stopped.
+maybe_drain_nodedown_msg(What, DownNode) when DownNode =/= node(), What =/= node ->
+    receive
+        {membership, {node, down, DownNode}} -> ok
+    after ?DRAIN_TIMEOUT -> ok
+    end;
+maybe_drain_nodedown_msg(_, _) ->
+    ok.
+
+locking_nodes() ->
+    locking_nodes(mria_rlog:role()).
+
+locking_nodes(core) ->
+    mria_mnesia:running_nodes();
+locking_nodes(replicant) ->
+    %% No reliable source to get the list of core nodes, unfortunately.
```

Comment:
maybe not critical to this PR. but it would be good to have mria brodcast core nodes to all replicants ?

Reply:
For sure, it has this mechanism, but I find it is not very suitable, especially when called upon app start in `init/1`, please see my comment: https://github.com/emqx/emqx/pull/12294#discussion_r1452553704

Reply:
The replicant node inserts info about core nodes by doing async pings, so at the point when emqx_router_helper is started, it can probably get an empty core nodes list, unless enough time has passed..

Reply:
@keynslug suggested a good alternative for now:
```
mria:async_dirty(?ROUTE_SHARD, fun mria_mnesia:running_nodes/0, []).
```